apps/studio/src/server/modules/rate-limit/rate-limit.service.ts

```diff
@@ -0,0 +1,52 @@
+import { type NextApiRequest } from "next"
+import { PrismaClient } from "@prisma/client"
+import { TRPCError } from "@trpc/server"
+import {
+  RateLimiterMemory,
+  RateLimiterPrisma,
+  RateLimiterRes,
+} from "rate-limiter-flexible"
+
+import { type RateLimitMetaOptions } from "./types"
+import { getRateLimitFingerprint } from "./utils"
+
+// Default 5 queries per second fallback
+export const rateLimiterMemory = new RateLimiterMemory({
+  points: 5,
+  duration: 1,
+})
+
+export async function checkRateLimit({
+  rateLimitOptions,
+  req,
+  prisma,
+}: {
+  rateLimitOptions: RateLimitMetaOptions
+  req: NextApiRequest
+  prisma: PrismaClient
+}) {
+  const max = rateLimitOptions.max ?? 5
+  const windowMs = rateLimitOptions.windowMs ?? 1000
+
+  const store = new RateLimiterPrisma({
+    storeClient: prisma,
+    points: max,
+    duration: windowMs / 1000, // in seconds
+    insuranceLimiter: rateLimiterMemory,
+  })
+
+  const fingerprint = getRateLimitFingerprint({ type: "IP", req })
```

Comment:
whats this fingerprint?

Reply:
Key to store in our table for this particular rate limit